### DISCUSSION TITLE
Why GGUF supports mmap specially?

### DISCUSSION THREAD

**Discussion by SanftMonster** [ggml-org/ggml#492] [Q&A]
According to [the doc of GGUF](https://github.com/philpax/ggml/blob/gguf-spec/docs/gguf.md), GGUF format has an advantage that it supports `mmap`, while `ggml` not. I'm so curious about it so I opened a discussion here. 

In my thought, `mmap` maps an area of file to an area of memory. As long as the file content we want to map is contiguous, it will work. According to [the code converting pth to ggml](https://github.com/ggerganov/ggml/blob/69bf842b39e6a53218eda35852a91d395d357e18/examples/sam/convert-pth-to-ggml.py#L144C22-L144C22), it writes the numpy array data to the file, which I think indicates that the data area in the file is contiguous. 

Maybe I mistaken something and I'll appreciate it if someone could correct me.😊

@philpax Please allow me to apologize for the interruption.😶‍🌫️ I think you are the most knowledgeable man about GGUF.

**Comment by Green-Sky** [comment]
It is not a difference, but a desired feature.

btw @philpax , here the `GGML` should be `GGUF` (i think):
> Extensible: new features can be added to GGML without breaking compatibility with existing models.

sidenote: the `GGJT` version of `GGML` referenced was specifically introduced to make it mmap-able.

**Reply by philpax** [reply]
Fixed. And yeah, this is right - the GGJT format supported mmap, we wanted to carry that forward into GGUF and make sure it was a primary design consideration.

**Reply by SanftMonster** [reply]
Thank you for all your answers!
